# TODO List

Progress | Process Name
:------------ | :-------------
&#9745; | FastQC |  
&#9745; | Fastqmcf |  
&#9745; | FastQC(trimmed) |  
&#9745; | Hisat2 (All genes) |  
&#9745; | Hisat2 (rRNA) |  
&#9745; | Hisat2 (SIRVome) |  
&#9745; | RSEM (All genes) |  
&#9745; | RSEM (SIRVome) |  
&#9745; | Bam to BigWig |  
&#9745; | adjust_bed_noncoding |  
&#9745; | RSeQC(read_distribution.py, infer_experiment.py, junction_annotation.py, inner_distance.py) |  
&#9745; | readcoverage.jl (All genes) |  
&#9745; | readcoverage.jl (SIRV genes) |  
&#9745; | FeatureCounts (All-genes GTF) |  
&#9745; | output TPMCounts |  
&#9745; | FeatureCounts (Mitochondrial GTF) |  
&#9745; | FeatureCounts (rRNA GTF)  |  
&#9745; | FeatureCounts (Histone GTF)  |  
&#9745; | edgeR MDS and heatmap |  
&#9745; | ERCC corr barplot |  
&#9745; | Assigned to genome rate barplot|  
&#9745; | FeatureCounts mapped rate barplot(allgene, mt, histone) |  
&#9745; | Create plots from TPM counts (num of detected genes, dimentianl reduction) |  
&#9745; | Create plots from RSEM TPM counts (num of detected genes, num of detected transcripts) |  
&#9745; | SIRV entropy barplot |  
&#9745; | MultiQC Report |  
&#9745; | helpMessage() |  
&#9745; | get_software_versions() |  
